### PR TITLE
Adds `runtime/compstore` runtime component store

### PR DESCRIPTION
**Pull request by JoshVanL**
compstore is a new runtime package which dynamically loads and unloads runtime components. It is used as a consistent store for runtime services to ensure they all have the same view of the world.

This is a predicate for [hot reloading](https://github.com/dapr/dapr/issues/1172), whereby all runtime services will need to use the same store of components loaded, which may be removed, added and updated, multiple times during runtime.

### CODE REVIEW

File: pkg/http/api.go [hunk 1]
```diff
@@ -76,17 +73,14 @@ type API interface {
 }
 
 type api struct {
-	universal                  *universalapi.UniversalAPI
+	*universalapi.UniversalAPI
```
Comment (ItalyPaleAle) [inline pkg/http/api.go:76]: Why this change? Not sure we should have `api` depend on this, since all methods would automatically be exported and if we have a method with the same name, but different arguments, there'd be a conflict IIRC

File: pkg/runtime/runtime.go [hunk 2]
```diff
@@ -913,10 +867,13 @@ func (a *DaprRuntime) subscribeTopic(parentCtx context.Context, name string, top
 		policyRunner := resiliency.NewRunner[any](ctx, policyDef)
 		_, err = policyRunner(func(ctx context.Context) (any, error) {
 			var pErr error
-			if a.runtimeConfig.ApplicationProtocol.IsHTTP() {
+			switch a.runtimeConfig.ApplicationProtocol {
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:870]: Please revert this and use `IsHTTP`, because we now have `https` and `h2c` too, and for gRPC (`!IsHTTP`) we have `grpcs`

File: pkg/runtime/runtime.go [hunk 3]
```diff
@@ -1264,12 +1224,12 @@ func (a *DaprRuntime) sendBindingEventToApp(bindingName string, data []byte, met
...
 	var appResponseBody []byte
-	path := a.inputBindingRoutes[bindingName]
-	if path == "" {
+	path, ok := a.compStore.GetInputBindingRoute(bindingName)
+	if !ok || path == "" {
 		path = bindingName
 	}
 
-	if !a.runtimeConfig.ApplicationProtocol.IsHTTP() {
+	if a.runtimeConfig.ApplicationProtocol == GRPCProtocol {
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1232]: Here too, please revert this

File: pkg/runtime/runtime.go [hunk 4]
```diff
@@ -1344,7 +1304,7 @@ func (a *DaprRuntime) sendBindingEventToApp(bindingName string, data []byte, met
 				}
 			}
 		}
-	} else {
+	} else if a.runtimeConfig.ApplicationProtocol == HTTPProtocol {
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1307]: Ditto

File: pkg/runtime/runtime.go [hunk 5]
```diff
@@ -1589,7 +1533,7 @@ func (a *DaprRuntime) getSubscribedBindingsGRPC() ([]string, error) {
 
 func (a *DaprRuntime) isAppSubscribedToBinding(binding string) (bool, error) {
 	// if gRPC, looks for the binding in the list of bindings returned from the app
-	if !a.runtimeConfig.ApplicationProtocol.IsHTTP() {
+	if a.runtimeConfig.ApplicationProtocol == GRPCProtocol {
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1536]: Ditto

File: pkg/runtime/runtime.go [hunk 6]
```diff
@@ -1602,9 +1546,9 @@ func (a *DaprRuntime) isAppSubscribedToBinding(binding string) (bool, error) {
 				return true, nil
 			}
 		}
-	} else {
+	} else if a.runtimeConfig.ApplicationProtocol == HTTPProtocol {
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1549]: Ditto

File: pkg/runtime/runtime.go [hunk 7]
```diff
@@ -1794,20 +1739,14 @@ func (a *DaprRuntime) initState(s componentsV1alpha1.Component) error {
...
-			actorStoreSpecified := false
-			for k, v := range props {
-				if strings.ToLower(k) == actorStateStore { //nolint:gocritic
-					actorStoreSpecified = utils.IsTruthy(v)
-				}
-			}
-
-			if actorStoreSpecified {
+			actorStoreSpecified := props[actorStateStore]
+			if actorStoreSpecified == "true" {
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1743]: Please revert this change too, which was made on purpose to make `actorstatestore` case-insensitive and match with IsTruthy

File: pkg/runtime/runtime.go [hunk 8]
```diff
@@ -1794,20 +1739,14 @@ func (a *DaprRuntime) initState(s componentsV1alpha1.Component) error {
...
 				a.actorStateStoreLock.Lock()
 				if a.actorStateStoreName == "" {
-					log.Info("Using '" + s.ObjectMeta.Name + "' as actor state store")
+					log.Infof("detected actor state store: %s", s.ObjectMeta.Name)
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1746]: Why were these logs reverted too?

File: pkg/runtime/runtime.go [hunk 9]
```diff
@@ -1871,9 +1806,9 @@ func (a *DaprRuntime) getSubscriptions() ([]runtimePubsub.Subscription, error) {
 	}
 
 	// handle app subscriptions
-	if a.runtimeConfig.ApplicationProtocol.IsHTTP() {
+	if a.runtimeConfig.ApplicationProtocol == HTTPProtocol {
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1809]: Here too, and same 2 lines below

File: pkg/runtime/runtime.go [hunk 10]
```diff
@@ -2962,12 +2887,13 @@ func (a *DaprRuntime) createAppChannel() (err error) {
 	}
 
 	var ch channel.AppChannel
-	if !a.runtimeConfig.ApplicationProtocol.IsHTTP() {
+	switch a.runtimeConfig.ApplicationProtocol {
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:2890]: Please don't change these too, same reason

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Overall LGTM but you've reverted some changes introduced in recent PRs that need to be kept please

File: pkg/http/api.go (hunk 1 above)
Comment (JoshVanL) [reply pkg/http/api.go:76]: This was to keep it consistent with `grpc`, but happy to revert.

File: pkg/http/api.go [hunk 11]
```diff
@@ -1815,7 +1793,7 @@ func (a *api) onGetActorState(reqCtx *fasthttp.RequestCtx) {
 		respond(reqCtx, withError(fasthttp.StatusInternalServerError, msg))
 		log.Debug(msg)
 	} else {
-		if resp == nil || len(resp.Data) == 0 {
+		if resp == nil || resp.Data == nil {
```
Comment (ItalyPaleAle) [inline pkg/http/api.go:1796]: Should not be reverted

```suggestion
		if resp == nil || len(resp.Data) == 0 {
```

**Comment by JoshVanL** [comment]
Thanks for the review @ItalyPaleAle. Apologies for the bad reverts- was a bad rebase+copy&paste PBKAC on my end.

File: pkg/runtime/runtime.go [hunk 12]
```diff
@@ -1264,8 +1213,8 @@ func (a *DaprRuntime) sendBindingEventToApp(bindingName string, data []byte, met
 	ctx, span := diag.StartInternalCallbackSpan(a.ctx, spanName, spanContext, a.globalConfig.Spec.TracingSpec)
 
 	var appResponseBody []byte
-	path := a.inputBindingRoutes[bindingName]
-	if path == "" {
+	path, ok := a.compStore.GetInputBindingRoute(bindingName)
+	if !ok || path == "" {
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1217]: Nit: if the property doesn't exist, it returns the zero value which is `""`. So checking `ok` is redundant 

```suggestion
	path := a.compStore.GetInputBindingRoute(bindingName)
	if path == "" {
```

**Review by ItalyPaleAle — COMMENTED** [review]
Overall LGTM but you've reverted some changes introduced in recent PRs that need to be kept please

File: pkg/http/api.go (hunk 11 above)
Comment (ItalyPaleAle) [reply pkg/http/api.go:1796]: Btw this is the reason why the E2E tests are failing I think. We made that change because the tests were not working properly.

**Review by ItalyPaleAle — DISMISSED** [review]

**Review by ItalyPaleAle — APPROVED** [review]

**Review by yaron2 — APPROVED** [review]
